### DISCUSSION TITLE
Testing Framework

### DISCUSSION THREAD

**Discussion by SirEndii** [cc-tweaked/CC-Tweaked#1732] [Q&A] [answered]
Greetings,

I'm currently *trying* to make game tests for AP. After learning how the whole semi-well documented game test stuff works, I tried to check the CC repository how CC tests their computers in game.

The things that immediately flew into my eyes were the api and core package in the testMod/java folder along with the TestAPI.java file
But these files aren't included in the CC maven artifact *for obvious reasons*. Shall I just copy these files over to my project, or is there a better way recommended?

And if I do that, is there anything I should keep in mind when I make my lua test scripts for my computers? Or even better, is there any documentation about creating tests for lua cc scripts that I couldn't find?

**Comment by SquidDev** [comment, marked as answer]
> Shall I just copy these files over to my project,

Yep! Everything should be licensed under the MPL 2.0, so you don't need to worry about CCPL restrictions.

> And if I do that, is there anything I should keep in mind when I make my lua test scripts for my computers? Or even better, is there any documentation about creating tests for lua cc scripts that I couldn't find?

[There's some notes in ARCHITECTURE.md about the game test code](https://github.com/cc-tweaked/CC-Tweaked/blob/mc-1.20.x/projects/ARCHITECTURE.md#game-tests), and I've tried to document most classes, but it's definitely not comprehensive.

Some vague notes which hopefully help a little:

 - Every computer is labelled with the test name ([example](https://github.com/cc-tweaked/CC-Tweaked/blob/a0f759527d79189d95e04a14d92f3b306cf44a3c/projects/common/src/testMod/resources/data/cctest/structures/computer_test.open_on_client.snbt#L42)). This then allows test to [find and interact](https://github.com/cc-tweaked/CC-Tweaked/blob/a0f759527d79189d95e04a14d92f3b306cf44a3c/projects/common/src/testMod/kotlin/dan200/computercraft/gametest/api/TestExtensions.kt#L76-L114) with the appropriate computer.
 - Most tests don't actually use Lua at all. Instead, we inject a custom `ILuaMachine` to allow running Kotlin code ([example](https://github.com/cc-tweaked/CC-Tweaked/blob/a0f759527d79189d95e04a14d92f3b306cf44a3c/projects/common/src/testMod/kotlin/dan200/computercraft/gametest/Turtle_Test.kt#L120-L123)). This makes tests much easier to write, as all the test code is in one place.
 - The computers which *are* running Lua code have id 0 (and so share a mount). That computer's [startup file](https://github.com/cc-tweaked/CC-Tweaked/blob/a0f759527d79189d95e04a14d92f3b306cf44a3c/projects/common/src/testMod/resources/data/cctest/computer/startup.lua) then finds the actual test code to run based on the label (and thus test name). Once the test has finished, it then uses `TestAPI` to mark the test as passing/failing, which is then picked up by the Java-side test.

**Reply by SirEndii** [reply]
Sweet sweet
Thank you very much for the answer
I will test that and implement some game tests tomorrow!
If I encounter any issues, I will reach back to you

**Reply by SirEndii** [reply]
Is it possible to make game tests for pocket computers?

**Reply by SquidDev** [reply]
Hrmr, in theory, but it'd be a bit more work - you'd need to have some entity ticking them.

**Reply by SirEndii** [reply]
Another thing
I'm currently doing what you've told me. I've copied the files from testMod/java successfully. But when I tried to add some tests like CC-T with kotlin, I noticed that It needs more than testMod/java provided. The [`testFixtures`](https://github.com/cc-tweaked/CC-Tweaked/tree/mc-1.19.2/src/testFixtures) source set as example

I am not entirely sure if that's the right way. And disclaimer, my experience with kotlin is very low 😅 

You can also track the changes in the PR https://github.com/IntelligenceModding/AdvancedPeripherals/pull/562, but that does currently not contain my changes I'm doing locally with the kotlin stuff

**Reply by SquidDev** [reply]
Ahh yes, you'll need `testFixtures` too. On the 1.20.1 branch, you'd probably need both `core` and `common`'s test fixtures.

**Reply by SirEndii** [reply]
Greetings @SquidDev 
I've worked on this the past 6 hours. I've implemented (copied) the needed stuff from the test source sets, but I have a major problem.

Everything compiles successfully, I also have auto-completion for the test dependencies like hamcrest or junit in testMod 
But when I run `runGameTestClient` or `runGameTestServer`, the test fail because they can't find any of the classes in the hamcrest or junit libraries. I have a test in the java module and the test from cc in the kotlin module, but all just throw on running `NoClassDefFoundError: org/junit/jupiter/api/Assertions` or  `NoClassDefFoundError: org/hamcrest/BaseMatcher`

I've tried so much, but I am unable to fix that issue, and I'm just confused right now. I think it's an issue with the task that it somehow uses the wrong classpath. But I am not sure and unable to test this

When I run `gradle dependencies` it shows me `org.junit.jupiter:junit-jupiter-api:5.10.2` in `testModRuntimeClasspath`


If you want to know what my setup is, all the changes are here → https://github.com/IntelligenceModding/AdvancedPeripherals/pull/562

**Reply by SquidDev** [reply]
This is due to BootstrapLauncher/ModLauncher being incredibly stupid and not obeying the classpath. Instead you need to add those files to the "Minecraft classpath".

In CC:T, we create a new `testMinecraftLibraries` configuration, and add all jars in that to the classpath:

https://github.com/cc-tweaked/CC-Tweaked/blob/de930c8d09a3a8560439ea43ab9a6e475b9fd03d/projects/forge/build.gradle.kts#L62-L74

**Reply by SirEndii** [reply]
Urgh
But thank you!
I will see how I can use this in a non kts environment

**Reply by SirEdvin** [reply]
Probably the most important reason why all my build scripts become kts :)

**Reply by SirEndii** [reply]
I first wanted to port our build scripts to kts
but I stopped right away

But the game tests now work. I just need to write the actual tests now

**Reply by SirEndii** [reply]
I have one more question @SquidDev 
How do you save your test structures? I am using the structure blocks, but it saves the structure as .nbt file, not .snbt while it seems that the cc structures are saved as snbt. 
When I try to run my test, the game wants to read a snbt file. `C:\Users\info\IdeaProjects\AdvancedPeripherals\src\testMod\resources\data\advancedperipheralstest\structures\peripheral_test.environment.snbt`. The path is 100% correctly, except for the snbt file ending 

I know that there are some ways using MC utilities, but I hope for something simpler

**Reply by SquidDev** [reply]
Initial disclaimer: I'm not actually sure what the Recommended workflow is — this bit of the gametest code predates Fabric/Forge support for gametests.

What CC:T does is set up an additional folder where game tests are loaded from. 

https://github.com/cc-tweaked/CC-Tweaked/blob/de930c8d09a3a8560439ea43ab9a6e475b9fd03d/projects/common/src/testMod/kotlin/dan200/computercraft/gametest/core/TestHooks.kt#L39

This allows you to use the `/test export` command, which saves the test structure as snbt instead.

**Reply by SirEndii** [reply]
EDIT:
It seems that we can just replace the world in the game test server folder with a flat world before we start the game test server or replace the server.properties


Greetings. We implemented a bunch of game tests which work flawlessly, and we build these tests on a flat world.
We have one issue, our geo scanner scans for specific ores. And we have a test for that, it just checks the amount of x ore in the structure. However, the actual tests seem to be run on a vanilla default world type world. So there are more ores generated (At least, this is what we noticed when we run these tests).
I know this isn't a CC specific question, but do you know how we could force a different world type or even load our own world for the tests?
